File: 2023/5.py
input = open(f"{__file__.replace('py', 'input')}").read().strip().split("\n\n")

seeds = [int(x) for x in input[0].split()[1::]]
maps = [
    [[int(y) for y in x.split(" ")] for x in map.split("\n")[1::]] for map in input[1::]
]

for i, seed in enumerate(seeds):
    for map in maps:
        for line in map:
            if line[1] <= seeds[i] < line[1] + line[2]:
                seeds[i] = line[0] + seeds[i] - line[1]
                break

print(min(seeds))


seeds = [int(x) for x in input[0].split()[1::]]
seed_ranges = []
for i in range(0, len(seeds), 2):
    seed_ranges.append([seeds[i], seeds[i + 1]])


def in_seeds(n):
    for seed in seed_ranges:
        if seed[0] <= n < seed[0] + seed[1]:
            return True
    return False


i = 100_000_000
while True:
    n = i
    for map in reversed(maps):
        for line in map:
            if line[0] <= n < line[0] + line[2]:
                n = line[1] + n - line[0]
                break

    if in_seeds(n):
        break
    i += 1

print(i)

File: 2023/5.rb
input = File.read(File.basename(__FILE__).gsub('rb', 'input'))
input = input.split("\n\n")

seeds = input[0].split(' ')[1..].map(&:to_i)
maps = input[1..].map { _1.split("\n")[1..].map { |e| e.split.map(&:to_i) } }

(0...seeds.size).each do |i|
  maps.each do |map|
    map.each do |line|
      if seeds[i] >= line[1] && seeds[i] < line[1] + line[2]
        seeds[i] = line[0] + seeds[i] - line[1]
        break
      end
    end
  end
end

p seeds.min


seeds = input[0].split(' ')[1..].map(&:to_i)
maps = maps.map { |map| map.map { _1 + [_1[0] + _1[2], _1[1] - _1[0]] } }

def in_seeds(seeds, n)
  seeds.each_slice(2) do |a, b|
    return true if n >= a && n < a + b
  end
  false
end

# runs in ~14 mins from 0
(100_000_000..).each do |i|
  n = i
  maps.reverse.each do |map|
    map.each do |line|
      if n >= line[0] && n < line[3]
        n += line[4]
        break
      end
    end
  end
  if in_seeds(seeds, n)
    p i
    break
  end
end
